### PR TITLE
Pods Egress DSCP QoS proposal

### PR DESCRIPTION
**Pull request by oribon**
Signed-off-by: Ori Braunshtein <obraunsh@redhat.com>

### CODE REVIEW

File: enhancements/network/egress-qos.md
Comment (danwinship) [inline enhancements/network/egress-qos.md:22]: "The resources" or "The objects". Don't say "The CRs" when it's not actually relevant that they're Custom. (People do that a lot, but they shouldn't.)

File: enhancements/network/egress-qos.md
Comment (danwinship) [inline enhancements/network/egress-qos.md:46]: Why per-namespace? Why not per-pod? Why allow a single PodQoS object to apply to multiple namespaces, rather than requiring a separate PodQoS for each namespace? Why per-destination-CIDR rather than per-destination-port or based on time of day or anything else?

Even if the current decisions solve the problem for the current customer, will they solve the problem for all customers?

File: enhancements/network/egress-qos.md
Comment (danwinship) [inline enhancements/network/egress-qos.md:53]: As currently defined, yes, because this object has no connection to the `default` namespace

File: enhancements/network/egress-qos.md
Comment (danwinship) [inline enhancements/network/egress-qos.md:67]: You should not need to match individual ports; you can match the address set for the namespace

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:46]: I went with per-namespace and not per-pod because that's the current requirements (which tbh are pretty vague right now), while per-pod(s) could be added later on.
Rethinking with your comments, we could have the CRD namespace scoped, with subsections corresponding to a CIDR/+port, leaving room for further granularity like podSelectors in the future, while enforcing 1 PodQos per namespace.
That way it might be less ambiguous in comparison to a cluster scoped CRD - it'll be clearer from where each Namespace gets its QoS.
Something like:
```yaml
kind: PodQos
apiVersion: k8s.ovn.org/v1
metadata:
  name: default
  namespace: ns1
spec:
  egress:
    - dstCIDR: 0.0.0.0/0
      port: 80
      dscp: 46
      priority: 1 # needed if overlapping destinations are used?
      # PodSelector could be another field here in the future?
    - dstCIDR: 1.2.3.4/32
      port: 443
      dscp: 32
      priority: 2
...
```
Is this approach better/cleaner?

IMO using dstCIDR/+port (or none) to determine when to apply DSCP makes sense, and tbh I'm not too familiar with other possible requirements to tell if there's anything better as a starting point or if it contradicts with another option.

File: enhancements/network/egress-qos.md
Comment (danwinship) [reply enhancements/network/egress-qos.md:21]: If I was going to nitpick the name, it would be the "Pod" part more than the "QoS" part... (Isn't it more like "NamespaceQos"? or "NamespaceEgressQos"?)

Maybe as we nail down the use case (and possible future changes to it) some more a better name will become clear.

Comment (tssurya) [reply enhancements/network/egress-qos.md:21]: I like "EgressQoS", just like we have egressIPs, egressFirewalls and maybe in the future egressGateways, fits into the picture since afterall it is QoS for egress traffic, restricting to Pod or Namespace at this point might hurt us in the future if we think of changing its scope?

**Review by tssurya — COMMENTED** [review]
/cc

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: > Like, if this object is _only_ going to be used to describe QoS-related stuff that is _only_ acted on by out-of-cluster components, then that suggests one design. But if we wanted to handle QoS-related stuff more generally, then that might suggest a different design, etc.

Well, I think our main (or maybe only) focus is the former. Going by that I updated the file to reflect that, pretty much following the `EgressFirewall` approach - having a namespace-scoped CRD `EgressQos`, affecting all of the pods in the resource's namespace. I think it makes sense in this context, and keeps this egress only - leaving room for maybe small additions like marking by destination port if we'd like.

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:23]: nit:
```suggestion
By introducing a new CRD `EgressQoS`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR.
```

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:49]:
```suggestion
To achieve egress DSCP marking on pods, we introduce a new CRD `EgressQoS`, which allows specifying a set of DSCP values and destination CIDRs - meaning traffic coming from pods on the namespace heading to each destination CIDR will be marked with the given DSCP value.
```

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:51]: Keep the proposal brief and move this yaml into impementation details section maybe?

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:66]: in OVN accordingly so that the necessary flows can be programmed?
nbctl is just a utility.

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:69]: again move the details in implementation section?

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:85]: I know we spoke about this offline, but we need to decide if its going to be on node switch or join switch depending on the gateway or if there isn't much perf impact for both modes we can probably leave it on the node switch.

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:61]: I wonder how we do it for firewall code today?

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:85]:
```suggestion
- Adding a controller in OVN-K that watches `EgressQoS` and `Pod` objects, which will create the relevant QoS objects in OVN and flows in OVS. These rules will be added on node local switches in the relevant nodes or something???
```

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:95]:
```suggestion
* IPv4/IPv6 E2E that validates egress traffic from a namespace is marked with the correct DSCP value by creating and deleting `EgressQoS`, setting up src pods and host-networked destination pods.
```

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:76]: We should fill this with the telco use cases?

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:99]: Mark the irrelevant sections with N/A

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:102]: This is Dev Preview in 4.11 right?
Tech Preview in 4.12 and GA in 4.13? Is that the goal?

**Review by tssurya — CHANGES_REQUESTED** [review]

File: enhancements/network/egress-qos.md
Comment (tssurya) [inline enhancements/network/egress-qos.md:10]: suggestion: you can make Tim approver for OVN-K side, Dan for API side, Ben Bennett who leads networking.

File: enhancements/network/egress-qos.md
Comment (mccv1r0) [inline enhancements/network/egress-qos.md:43]: So the external world hurries up to get a packet to OCP that we treat as best effort?

File: enhancements/network/egress-qos.md
Comment (dcbw) [inline enhancements/network/egress-qos.md:81]: We need to address IPv6 too. When you add examples, could you add IPv6 ones too?

File: enhancements/network/egress-qos.md
Comment (msherif1234) [reply enhancements/network/egress-qos.md:23]: sorry trying to understand this feature, ovnk will mark and the some entity outside the cluster will check DSCP and take some actions so we are talking about service traffic right so CIDR here are ClusterIP or LBServiceIP right ?

File: enhancements/network/egress-qos.md
Comment (msherif1234) [reply enhancements/network/egress-qos.md:61]: how can u refelect the priority setting in the packet header ? setting dscp should be enough IMO

File: enhancements/network/egress-qos.md
Comment (msherif1234) [inline enhancements/network/egress-qos.md:45]: Hmm who inside the cluster will be looking at the marked header ? unless we have ingress ACL that filter on dscp and take actions but this seems to be irrelevant to this proposal ?

File: enhancements/network/egress-qos.md
Comment (msherif1234) [reply enhancements/network/egress-qos.md:69]: we should also be clear what we will be using for from-lport I assum QoS has the same limitation like ACL and it only applies to switch's lp so which switch port will be used here nodeSwitch or JoinSwitch I assume its the nodeSwitch but that needs to be called out in the implementation detail section.

File: enhancements/network/egress-qos.md
Comment (msherif1234) [reply enhancements/network/egress-qos.md:85]: oh @tssurya I was thinking about the same exact thing before I read your comment, since it seems we will move the egress ACL to nodeSwitch then it seems like having QoS there also will make sense ?

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: Yeah some entity outside of the cluster will take actions based on the dscp value, but the CIDRs here are these entities - destinations that are external.
The purpose is to mark egress traffic from specific pods based on the relevant rule in the namespace's EgressQoS: by dstCIDR and dscp value.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: done

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:59]: done

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: whoops, fixed

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: > I wonder how we do it for firewall code today?

iiuc it relies on the rules to be in a descending order: https://github.com/ovn-org/ovn-kubernetes/blob/a24a233130fb87e07557d8c1579ad52a861fde96/go-controller/pkg/ovn/egressfirewall.go#L341
We can do that here and/or maybe sorting the rules first by mask/prefix length?


> how can u refelect the priority setting in the packet header ? setting dscp should be enough IMO

this priority relates to the one in the QoS table: https://man7.org/linux/man-pages/man5/ovn-nb.5.html#QoS_TABLE , iiuc we should differentiate similar matches using this.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:43]: Not sure I understand, but this PR concerns only marking egress traffic from specified pods. We won't do anything from Openshift's side for incoming traffic regarding DSCP related/unrelated to what's added here.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:45]: This is a non-goal, thought it's worth writing it here that we don't care about any traffic other than traffic that exits the cluster. Do you think I should remove this point?

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: sure added a short one for now

File: enhancements/network/egress-qos.md
Comment (abhat) [inline enhancements/network/egress-qos.md:23]: Should the API design also allow for ingress specification, or would that be a separate CRD?

File: enhancements/network/egress-qos.md
Comment (abhat) [inline enhancements/network/egress-qos.md:24]: The API should probably allow for namespace selector or pod selectors to allow for a more granular control or do we want this to be purely a per-namespace thing?

File: enhancements/network/egress-qos.md
Comment (abhat) [inline enhancements/network/egress-qos.md:139]: s/Dscp/DSCP?

File: enhancements/network/egress-qos.md
Comment (abhat) [inline enhancements/network/egress-qos.md:142]: Should this be optional? In other words, if the user wants all egress traffic marked with the same QoS value, there should be no need to specify 0.0.0.0/0

Comment (abhat) [reply enhancements/network/egress-qos.md:142]: Need kubebuilder validation for addresses and valid dscp code values.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:23]: In the current design Ingress would have to be in a separate CRD https://github.com/openshift/enhancements/pull/1035#discussion_r812840859

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:24]: I think it would be clearer having the CRD namespace-scoped and maybe a bit easier to implement that way. We'll achieve per-namespace egress dscp both ways, but won't have to manage/watch namespaces.
I believe pod selectors is a field we could add later as part of the egress rule if we'd want.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: done

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:142]: > Should this be optional? In other words, if the user wants all egress traffic marked with the same QoS value, there should be no need to specify 0.0.0.0/0

I don't have a strong opinion about this, do you think defaulting to 0.0.0.0/0 or ::/0 is clearer?



> Need kubebuilder validation for addresses and valid dscp code values.

For the dscp I added one, for addresses I'm a bit hesitant about writing a good enough regex that covers all IPv4/6 CIDRs. In any case, we'll parse the addresses with `net.ParseCIDR`.

Comment (abhat) [reply enhancements/network/egress-qos.md:142]: Check how the service CIDR and cluster IP subnets are validated in network.config.openshift.io object in `openshift/api`.

Comment (abhat) [reply enhancements/network/egress-qos.md:142]: If the 80% users want all traffic in the namespace to be marked a certain way, then the dst cidr field seems like an overhead. OTOH, if specifying only a subset of destinations for which the packets should be marked with a certain DSCP value is the more common case, then I would lean on keeping the field mandatory.

File: enhancements/network/egress-qos.md
Comment (trozet) [reply enhancements/network/egress-qos.md:74]: If we only care about egress traffic, then we should be implementing this at the join switch. Otherwise we are going to marking east west packets for no reason.

File: enhancements/network/egress-qos.md
Comment (trozet) [inline enhancements/network/egress-qos.md:68]: what does 0.0.0.0/0 represent in OVN? Is that the equivalent of "any"? If so what happens with overlapping CIDRs?

File: enhancements/network/egress-qos.md
Comment (trozet) [inline enhancements/network/egress-qos.md:70]: why do we only match on destination CIDR and then assume the source is all addresses in this namespace? I think it would be better to add srcCIDR option as well so that only certain pods may be selected.

File: enhancements/network/egress-qos.md
Comment (trozet) [inline enhancements/network/egress-qos.md:60]: I presume in the future that:
"Consolidating with current `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` annotations." will end up being consolidated into a QOS CRD. In that case I suspect the relevant field, like egress-bandwidth would be added under this CRD type. Do we want to end up having 2 separate EgressQoS and IngressQoS CRDs, or should we just have one QoS CRD?

File: enhancements/network/egress-qos.md
Comment (trozet) [inline enhancements/network/egress-qos.md:75]: note this "2" is priority:
[--may-exist] qos-add switch direction priority match [dscp=dscp]

how will the CRD determine the priority to use for these?

**Comment by trozet** [comment]
@oribon can you resolve some of the resolved comments so it cleans up this review a bit?

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:75]: I thought following what egressFirewall is doing (https://github.com/ovn-org/ovn-kubernetes/blob/a24a233130fb87e07557d8c1579ad52a861fde96/go-controller/pkg/ovn/egressfirewall.go#L341) - relying on the rules of the CRD to be in a descending order priority-wise. The user would have to put them in an order that'll work like having a specific CIDR in a higher priority than a broader one that includes it.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:70]: yes that would be "any". we're still not sure if we should allow the field to be empty and that'll default it to 0.0.0.0/0 or ::/0
regarding overlapping CIDRs that'll need to be handled by the priority setting: https://github.com/openshift/enhancements/pull/1035#discussion_r825411903

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:72]: Wouldn't srcCIDR be a bit complicated for the user? Adding a PodSelector per rule is possible with the current design but I left it out for now as it seems that it is not a requirement.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:62]: I think that we want to keep this CRD focused on egress QoS, so if we do decide to have some ingress QoS stuff it'll have to be under a different CRD.
https://github.com/openshift/enhancements/pull/1035#discussion_r811211259

**Comment by oribon** [comment]
> @oribon can you resolve some of the resolved comments so it cleans up this review a bit?

done

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md:76]: done, added some words about LGW vs SGW

File: enhancements/network/egress-qos.md
Comment (abhat) [reply enhancements/network/egress-qos.md:77]: that API is very confusing to me. do we want to explore an explicit priority field?

File: enhancements/network/egress-qos.md
Comment (abhat) [reply enhancements/network/egress-qos.md:72]: Actually revisiting my conversation with the customer, pod selector is very much a requirement. We want to be able to say `Pod with X label gets DSCP marking of Y for all of its traffic` as a functional MVP. namespace selector can come later perhaps and if we are making the CRD namespace scoped, then we will be OK.

Comment (oribon) [reply enhancements/network/egress-qos.md:78]: Added a PodSelector field for the QoS rules, and a bit about how we can use AddressSets to make sure matching pods are marked with the right DSCP.

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: we could, but I'm not sure how useful it could be for the user.
with the rules for some namespace being something like :
```
  - dscp: 46
    dstCIDR: 1.2.3.4/32
  - dscp: 40
    dstCIDR: 1.2.3.0/24
  - dscp: 30
    dstCIDR: 0.0.0.0/0
```
would it add any functionality to let the user set some priorities here? would it make any difference if he sets them 3,2,1 or 80,75,5?
looking at a larger set of rules, is it simpler to observe the priority of each rule to understand what takes precedence or rather relying on the order (or some other option)?

Comment (trozet) [reply enhancements/network/egress-qos.md]: I think we should just make it work like egress firewall

File: enhancements/network/egress-qos.md
Comment (abhat) [inline enhancements/network/egress-qos.md:83]: What does the `priority: Critical` mean in the context of the `podSelector`?

Comment (abhat) [reply enhancements/network/egress-qos.md:83]: Oh it is just an example of matching pods with labels as critical?

Comment (oribon) [reply enhancements/network/egress-qos.md:83]: yes :sweat_smile:

File: enhancements/network/egress-qos.md
Comment (oribon) [reply enhancements/network/egress-qos.md]: after also talking with Tim and Dan this field should be optional, updated.
thanks :)

**Review by trozet — APPROVED** [review]
/approve

**Comment by abhat** [comment]
/hold
Need additional API review.

File: enhancements/network/egress-qos.md
Comment (trozet) [inline enhancements/network/egress-qos.md:61]: do we need to put it on 2 different switches (depending on gateway mode)? It would be easier to just do this on the node switch for both modes I think. Also, is the qos evaluation done after load balancer DNAT so that your dst match would be post service DNAT?

Comment (trozet) [reply enhancements/network/egress-qos.md:61]: I know I previously made a comment that steered you towards splitting to worker and join switch for different modes, but I think since we match on dest ip its OK to do them both on the worker switch.

@dceara pointed out that the qos eval is done pre-LB DNAT, so matching on destinations that are k8s endpoints woudln't work. Maybe its worth just noting that here?

Comment (dceara) [reply enhancements/network/egress-qos.md:61]: Is there any reason why the traffic can't be dscp-marked in the "egress" pipeline (egress from OVN perspective)?  If we use "to-lport" QOS rules then there's no issue, those are applied after LB DNAT, so they could match on endpoint IPs.

Comment (oribon) [reply enhancements/network/egress-qos.md]: > I know I previously made a comment that steered you towards splitting to worker and join switch for different modes, but I think since we match on dest ip its OK to do them both on the worker switch.
> 
> @dceara pointed out that the qos eval is done pre-LB DNAT, so matching on destinations that are k8s endpoints woudln't work. Maybe its worth just noting that here?

changed it to use node switches regardless of sgw/lgw.



> Is there any reason why the traffic can't be dscp-marked in the "egress" pipeline (egress from OVN perspective)? If we use "to-lport" QOS rules then there's no issue, those are applied after LB DNAT, so they could match on endpoint IPs.

no reason, changed it to to-lport as it seems that it still works

**Pull request merged**
